Some configuration change crashes the app

### Steps to reproduce

1. Open app
2. Trigger a configuration changes which is not among the ignored ones in the manifest: `android:configChanges="orientation|screenSize|screenLayout|keyboardHidden|uiMode"` For instance go to "Wallpaper & Style" and change the wallpaper colors.
3. Reopen the app.
4. App crashes with:

```
Thread: main, Exception: java.lang.RuntimeException: Unable to resume activity {io.element.android.x.debug/io.element.android.x.MainActivity}: kotlin.UninitializedPropertyAccessException: lateinit property mainNode has not been initialized
                                                                                                    	at android.app.ActivityThread.performResumeActivity(ActivityThread.java:4773)
                                                                                                    	at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:4806)
                                                                                                    	at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:54)
                                                                                                    	at android.app.servertransaction.ActivityTransactionItem.execute(ActivityTransactionItem.java:45)
                                                                                                    	at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176)
                                                                                                    	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
                                                                                                    	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2307)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:106)
                                                                                                    	at android.os.Looper.loopOnce(Looper.java:201)
                                                                                                    	at android.os.Looper.loop(Looper.java:288)
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:7872)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
                                                                                                    Caused by: kotlin.UninitializedPropertyAccessException: lateinit property mainNode has not been initialized
                                                                                                    	at io.element.android.x.MainActivity.getMainNode(MainActivity.kt:42)
                                                                                                    	at io.element.android.x.MainActivity.onNewIntent(MainActivity.kt:87)
                                                                                                    	at android.app.Activity.performNewIntent(Activity.java:8321)
                                                                                                    	at android.app.Instrumentation.callActivityOnNewIntent(Instrumentation.java:1517)
                                                                                                    	at android.app.Instrumentation.callActivityOnNewIntent(Instrumentation.java:1530)
                                                                                                    	at android.app.ActivityThread.deliverNewIntents(ActivityThread.java:3822)
                                                                                                    	at android.app.ActivityThread.performResumeActivity(ActivityThread.java:4756)
                                                                                                    	at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:4806) 
                                                                                                    	at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:54) 
                                                                                                    	at android.app.servertransaction.ActivityTransactionItem.execute(ActivityTransactionItem.java:45) 
                                                                                                    	at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176) 
                                                                                                    	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97) 
                                                                                                    	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2307) 
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:106) 
                                                                                                    	at android.os.Looper.loopOnce(Looper.java:201) 
                                                                                                    	at android.os.Looper.loop(Looper.java:288) 
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:7872) 
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method) 
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548) 
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936) 
```

I'm not sure this is actually due to the configuration change. The issue seems more about `onNewIntent()` (which I guess is called because we are opening a singleInstance existing activity by clicking on its launcher icon) is racing with `onCreate()` for accessing `var lateinit mainNode` .

This opens another question? Why MainActivity has `android:launchMode="singleInstance"` ? Removing it actually fixes the issue, but I was wondering what is the particular reason that forces us to use `singleInstance`. 


### Outcome

#### What did you expect?
App crashed

#### What happened instead?
App did not crash

### Your phone model

Pixel 6a or Emulator

### Operating system version

Android 13

### Application version and app store

EAX latest from develop

### Homeserver

matrix.org

### Will you send logs?

Yes

### Are you willing to provide a PR?

Yes

--- a/app/src/main/kotlin/io/element/android/x/MainActivity.kt
+++ b/app/src/main/kotlin/io/element/android/x/MainActivity.kt
@@ -23,6 +23,7 @@ import androidx.compose.foundation.background
 import androidx.compose.foundation.layout.Box
 import androidx.compose.foundation.layout.fillMaxSize
 import androidx.compose.material3.MaterialTheme
+import androidx.compose.runtime.Composable
 import androidx.compose.ui.Modifier
 import androidx.core.splashscreen.SplashScreen.Companion.installSplashScreen
 import androidx.core.view.WindowCompat
@@ -39,7 +40,7 @@ private val loggerTag = LoggerTag("MainActivity")
 
 class MainActivity : NodeComponentActivity() {
 
-    lateinit var mainNode: MainNode
+    private lateinit var mainNode: MainNode
 
     override fun onCreate(savedInstanceState: Bundle?) {
         Timber.tag(loggerTag.value).w("onCreate, with savedInstanceState: ${savedInstanceState != null}")
@@ -49,26 +50,32 @@ class MainActivity : NodeComponentActivity() {
         appBindings.matrixClientsHolder().restore(savedInstanceState)
         WindowCompat.setDecorFitsSystemWindows(window, false)
         setContent {
-            ElementTheme {
-                Box(
-                    modifier = Modifier
-                        .fillMaxSize()
-                        .background(MaterialTheme.colorScheme.background),
-                ) {
-                    NodeHost(integrationPoint = appyxIntegrationPoint) {
-                        MainNode(
-                            it,
-                            appBindings.mainDaggerComponentOwner(),
-                            plugins = listOf(
-                                object : NodeReadyObserver<MainNode> {
-                                    override fun init(node: MainNode) {
-                                        mainNode = node
-                                        mainNode.handleIntent(intent)
-                                    }
+            MainContent(appBindings)
+        }
+    }
+
+    @Composable
+    private fun MainContent(appBindings: AppBindings) {
+        ElementTheme {
+            Box(
+                modifier = Modifier
+                    .fillMaxSize()
+                    .background(MaterialTheme.colorScheme.background),
+            ) {
+                NodeHost(integrationPoint = appyxIntegrationPoint) {
+                    MainNode(
+                        it,
+                        appBindings.mainDaggerComponentOwner(),
+                        plugins = listOf(
+                            object : NodeReadyObserver<MainNode> {
+                                override fun init(node: MainNode) {
+                                    Timber.tag(loggerTag.value).w("onMainNodeInit")
+                                    mainNode = node
+                                    mainNode.handleIntent(intent)
                                 }
-                            )
+                            }
                         )
-                    }
+                    )
                 }
             }
         }
@@ -80,11 +87,17 @@ class MainActivity : NodeComponentActivity() {
      * - a notification is clicked.
      * - the app is going to background (<- this is strange)
      */
-    override fun onNewIntent(intent: Intent?) {
+    override fun onNewIntent(intent: Intent) {
         super.onNewIntent(intent)
         Timber.tag(loggerTag.value).w("onNewIntent")
-        intent ?: return
-        mainNode.handleIntent(intent)
+        // If the mainNode is not init yet, keep the intent for later.
+        // It can happen when the activity is killed by the system. The methods are called in this order :
+        // onCreate(savedInstanceState=true) -> onNewIntent -> onResume -> onMainNodeInit
+        if (::mainNode.isInitialized) {
+            mainNode.handleIntent(intent)
+        } else {
+            setIntent(intent)
+        }
     }
 
     override fun onPause() {

--- a/app/src/main/kotlin/io/element/android/x/MainNode.kt
+++ b/app/src/main/kotlin/io/element/android/x/MainNode.kt
@@ -68,13 +68,13 @@ class MainNode(
     }
 
     private val roomFlowNodeCallback = object : RoomFlowNode.LifecycleCallback {
-        override fun onFlowCreated(room: MatrixRoom) {
+        override fun onFlowCreated(owner: String, room: MatrixRoom) {
             val component = bindings<RoomComponent.ParentBindings>().roomComponentBuilder().room(room).build()
-            mainDaggerComponentOwner.addComponent(room.roomId.value, component)
+            mainDaggerComponentOwner.addComponent(owner, component)
         }
 
-        override fun onFlowReleased(room: MatrixRoom) {
-            mainDaggerComponentOwner.removeComponent(room.roomId.value)
+        override fun onFlowReleased(owner: String, room: MatrixRoom) {
+            mainDaggerComponentOwner.removeComponent(owner)
         }
     }
 

--- a/app/src/main/kotlin/io/element/android/x/MainNode.kt
+++ b/app/src/main/kotlin/io/element/android/x/MainNode.kt
@@ -57,24 +57,24 @@ class MainNode(
     DaggerComponentOwner by mainDaggerComponentOwner {
 
     private val loggedInFlowNodeCallback = object : LoggedInFlowNode.LifecycleCallback {
-        override fun onFlowCreated(client: MatrixClient) {
+        override fun onFlowCreated(identifier: String, client: MatrixClient) {
             val component = bindings<SessionComponent.ParentBindings>().sessionComponentBuilder().client(client).build()
-            mainDaggerComponentOwner.addComponent(client.sessionId.value, component)
+            mainDaggerComponentOwner.addComponent(identifier, component)
         }
 
-        override fun onFlowReleased(client: MatrixClient) {
-            mainDaggerComponentOwner.removeComponent(client.sessionId.value)
+        override fun onFlowReleased(identifier: String, client: MatrixClient) {
+            mainDaggerComponentOwner.removeComponent(identifier)
         }
     }
 
     private val roomFlowNodeCallback = object : RoomFlowNode.LifecycleCallback {
-        override fun onFlowCreated(owner: String, room: MatrixRoom) {
+        override fun onFlowCreated(identifier: String, room: MatrixRoom) {
             val component = bindings<RoomComponent.ParentBindings>().roomComponentBuilder().room(room).build()
-            mainDaggerComponentOwner.addComponent(owner, component)
+            mainDaggerComponentOwner.addComponent(identifier, component)
         }
 
-        override fun onFlowReleased(owner: String, room: MatrixRoom) {
-            mainDaggerComponentOwner.removeComponent(owner)
+        override fun onFlowReleased(identifier: String, room: MatrixRoom) {
+            mainDaggerComponentOwner.removeComponent(identifier)
         }
     }
 

--- a/app/src/main/kotlin/io/element/android/x/di/MainDaggerComponentsOwner.kt
+++ b/app/src/main/kotlin/io/element/android/x/di/MainDaggerComponentsOwner.kt
@@ -38,6 +38,10 @@ class MainDaggerComponentsOwner @Inject constructor(@ApplicationContext context:
         daggerComponents.remove(identifier)
     }
 
+    /**
+     * We expose the dagger components in the opposite order they arrived.
+     * So we pick the most recent component when searching with the [io.element.android.libraries.architecture.bindings] methods.
+     */
     override val daggerComponent: Any
         get() = daggerComponents.values.reversed()
 }

--- a/appnav/src/main/kotlin/io/element/android/appnav/LoggedInFlowNode.kt
+++ b/appnav/src/main/kotlin/io/element/android/appnav/LoggedInFlowNode.kt
@@ -64,7 +64,6 @@ import io.element.android.services.analytics.api.AnalyticsService
 import io.element.android.services.appnavstate.api.AppNavigationStateService
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.flow.distinctUntilChanged
-import kotlinx.coroutines.flow.flowOf
 import kotlinx.coroutines.flow.launchIn
 import kotlinx.coroutines.flow.onEach
 import kotlinx.parcelize.Parcelize
@@ -118,9 +117,9 @@ class LoggedInFlowNode @AssistedInject constructor(
     }
 
     interface LifecycleCallback : NodeLifecycleCallback {
-        fun onFlowCreated(client: MatrixClient) = Unit
+        fun onFlowCreated(identifier: String, client: MatrixClient) = Unit
 
-        fun onFlowReleased(client: MatrixClient) = Unit
+        fun onFlowReleased(identifier: String, client: MatrixClient) = Unit
     }
 
     data class Inputs(
@@ -139,7 +138,7 @@ class LoggedInFlowNode @AssistedInject constructor(
         observeAnalyticsState()
         lifecycle.subscribe(
             onCreate = {
-                plugins<LifecycleCallback>().forEach { it.onFlowCreated(inputs.matrixClient) }
+                plugins<LifecycleCallback>().forEach { it.onFlowCreated(id, inputs.matrixClient) }
                 val imageLoaderFactory = bindings<MatrixUIBindings>().loggedInImageLoaderFactory()
                 Coil.setImageLoader(imageLoaderFactory)
                 inputs.matrixClient.startSync()
@@ -151,7 +150,7 @@ class LoggedInFlowNode @AssistedInject constructor(
             onDestroy = {
                 val imageLoaderFactory = bindings<MatrixUIBindings>().notLoggedInImageLoaderFactory()
                 Coil.setImageLoader(imageLoaderFactory)
-                plugins<LifecycleCallback>().forEach { it.onFlowReleased(inputs.matrixClient) }
+                plugins<LifecycleCallback>().forEach { it.onFlowReleased(id, inputs.matrixClient) }
                 appNavigationStateService.onLeavingSpace(id)
                 appNavigationStateService.onLeavingSession(id)
                 loggedInFlowProcessor.stopObserving()

--- a/appnav/src/main/kotlin/io/element/android/appnav/RoomFlowNode.kt
+++ b/appnav/src/main/kotlin/io/element/android/appnav/RoomFlowNode.kt
@@ -68,8 +68,8 @@ class RoomFlowNode @AssistedInject constructor(
 ) {
 
     interface LifecycleCallback : NodeLifecycleCallback {
-        fun onFlowCreated(room: MatrixRoom) = Unit
-        fun onFlowReleased(room: MatrixRoom) = Unit
+        fun onFlowCreated(owner: String, room: MatrixRoom) = Unit
+        fun onFlowReleased(owner: String, room: MatrixRoom) = Unit
     }
 
     data class Inputs(
@@ -83,14 +83,14 @@ class RoomFlowNode @AssistedInject constructor(
         lifecycle.subscribe(
             onCreate = {
                 Timber.v("OnCreate")
-                plugins<LifecycleCallback>().forEach { it.onFlowCreated(inputs.room) }
+                plugins<LifecycleCallback>().forEach { it.onFlowCreated(id, inputs.room) }
                 appNavigationStateService.onNavigateToRoom(id, inputs.room.roomId)
                 fetchRoomMembers()
             },
             onDestroy = {
                 Timber.v("OnDestroy")
                 inputs.room.close()
-                plugins<LifecycleCallback>().forEach { it.onFlowReleased(inputs.room) }
+                plugins<LifecycleCallback>().forEach { it.onFlowReleased(id, inputs.room) }
                 appNavigationStateService.onLeavingRoom(id)
             }
         )

--- a/appnav/src/main/kotlin/io/element/android/appnav/RoomFlowNode.kt
+++ b/appnav/src/main/kotlin/io/element/android/appnav/RoomFlowNode.kt
@@ -68,8 +68,8 @@ class RoomFlowNode @AssistedInject constructor(
 ) {
 
     interface LifecycleCallback : NodeLifecycleCallback {
-        fun onFlowCreated(owner: String, room: MatrixRoom) = Unit
-        fun onFlowReleased(owner: String, room: MatrixRoom) = Unit
+        fun onFlowCreated(identifier: String, room: MatrixRoom) = Unit
+        fun onFlowReleased(identifier: String, room: MatrixRoom) = Unit
     }
 
     data class Inputs(

